test/e2e/suite/issuers/ca/certificaterequest.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ca
+
+import (
+	"crypto/x509"
+	"net"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/test/e2e/framework"
+	"github.com/jetstack/cert-manager/test/e2e/util"
+)
+
+var _ = framework.CertManagerDescribe("CA Certificate", func() {
+	f := framework.NewDefaultFramework("create-ca-certificate")
+	h := f.Helper()
+
+	issuerName := "test-ca-issuer"
+	issuerSecretName := "ca-issuer-signing-keypair"
+	certificateRequestName := "test-ca-certificaterequest"
+	//certificateRequestSecretName := "test-ca-certificaterequest"
+
+	exampleDNSNames := []string{"dnsName1.co", "dnsName2.ninja"}
+	exampleIPAddresses := []net.IP{
+		[]byte{8, 8, 8, 8},
+		[]byte{1, 1, 1, 1},
+	}
+	exampleURIs := []string{"spiffe://foo.foo.example.net", "spiffe://foo.bar.example.net"}
+
+	JustBeforeEach(func() {
+		By("Creating an Issuer")
+		_, err := f.CertManagerClientSet.CertmanagerV1alpha1().Issuers(f.Namespace.Name).Create(util.NewCertManagerCAIssuer(issuerName, issuerSecretName))
+		Expect(err).NotTo(HaveOccurred())
+		By("Waiting for Issuer to become Ready")
+		err = util.WaitForIssuerCondition(f.CertManagerClientSet.CertmanagerV1alpha1().Issuers(f.Namespace.Name),
+			issuerName,
+			v1alpha1.IssuerCondition{
+				Type:   v1alpha1.IssuerConditionReady,
+				Status: v1alpha1.ConditionTrue,
+			})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		By("Cleaning up")
+		f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Delete(issuerSecretName, nil)
+		f.CertManagerClientSet.CertmanagerV1alpha1().Issuers(f.Namespace.Name).Delete(issuerName, nil)
+	})
+
+	Context("when the CA is the root", func() {
+		BeforeEach(func() {
+			By("Creating a signing keypair fixture")
+			_, err := f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Create(newSigningKeypairSecret(issuerSecretName))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should generate a valid certificate from CSR", func() {
+			certRequestClient := f.CertManagerClientSet.CertmanagerV1alpha1().CertificateRequests(f.Namespace.Name)
+
+			By("Creating a CertificateRequest")
+			cr, key, err := util.NewCertManagerBasicCertificateRequest(certificateRequestName, issuerName, v1alpha1.IssuerKind,
```

Comment:
nit: it'd be better to use the `gen` package pattern here for building resources as it makes it clearer what the structure of the resource we're creating is when glancing over the tests.

Reply:
Not sure what you mean here. Do you mean replacing all instances of `NewCertManagerBasicCertificateRequest` ?

Reply:
Yep - with something more like:

```
gen.CertificateRequest("cert-request-name",
    gen.SetCertificateRequestCSR(data),
    gen.SetCertificateRequestNotAfter(notAfter),
)
```

The logic for actually constructing a CSR can probably be better 'hidden' too, with another function that accepts fields to set on the CSR + returns a []byte which is the actual pem encoded x509 csr

Reply:
Doesn't have to be in this PR if you'd rather avoid it, but making that logic reusable/flexible will make it easier to write new test cases in future 😄